usr.bin/ul/ul.1
.\"	$NetBSD: ul.1,v 1.9 2003/08/07 11:16:52 agc Exp $
.\"
.\" Copyright (c) 1980, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)ul.1	8.1 (Berkeley) 6/6/93
.\"
.Dd June 6, 1993
.Dt UL 1
.Os
.Sh NAME
.Nm ul
.Nd do underlining
.Sh SYNOPSIS
.Nm
.Op Fl i
.Op Fl s
.Op Fl t Ar terminal
.Op Ar name Ar ...
.Sh DESCRIPTION
.Nm
reads the named files (or standard input if none are given)
and translates occurrences of underscores to the sequence
which indicates underlining for the terminal in use, as specified
by the environment variable
.Ev TERM .
The file
.Pa /usr/share/misc/termcap
is read to determine the appropriate sequences for underlining.
If the terminal is incapable of underlining, but is capable of
a standout mode then that is used instead.
If the terminal can overstrike,
or handles underlining automatically,
.Nm
degenerates to
.Xr cat 1 .
If the terminal cannot underline, underlining is ignored.
.Pp
The following options are available:
.Bl -tag -width Ds
.It Fl i
Underlining is indicated by a separate line containing appropriate
dashes `\-'; this is useful when you want to look at the underlining
which is present in an
.Xr nroff 1
output stream on a crt-terminal.
.It Fl s
By default, a space is handled just like any other character: a space
overstruck with an underscore produces an underlined space.  This is
logically consistent and produces the most sensible results when used
with terminals with a standout mode, especially as compared with what
programs such as
.Xr less 1
do.  Historically, such combinations have turned into single
underscores.
.Fl s
produces the historical behaviour.
.It Fl t Ar terminal
Overrides the terminal type specified in the environment with
.Ar terminal .
.El
.Sh ENVIRONMENT
The following environment variable is used:
.Bl -tag -width TERM
.It Ev TERM
The
.Ev TERM
variable is used to relate a tty device
with its device capability description (see
.Xr termcap 5 ) .
.Ev TERM
is set at login time, either by the default terminal type
specified in
.Pa /etc/ttys
or as set during the login process by the user in their
.Pa login
file (see for example
.Xr csh 1 Ns 's Ic setenv ) .
.El
.Sh SEE ALSO
.Xr colcrt 1 ,
.Xr man 1 ,
.Xr nroff 1
.Sh HISTORY
The
.Nm
command appeared in
.Bx 3.0 .
.Sh BUGS
.Xr nroff 1
usually outputs a series of backspaces and underlines intermixed
with the text to indicate underlining.  No attempt is made to optimize
the backward motion.


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision: 1.1142.2.10 $>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------


Changes from NetBSD 4.0 to NetBSD 5.0:
	evbppc: Add support for Xilinx Virtex II-Pro/4-FX. [freza 20061202]
	mac68k: Add support for four more NuBus video cards, including the
		VillageTronic MacPicasso 340. [hauke 20061206]
	puc(4): Add support for Perle Systems 4 and 8 ports modem boards.
		[manu 20071209]
	arm: Add support for NPE Ethernet on IXP425. [scw 20061210]
	arm: Add support for IXP425 watchdog timer. [scw 20061210]
	libutil: Add flags_to_string(3) and string_to_flags(3),
		moved from ls(1). [christos 20061214]
	arm: Add support for i2c, i2s, pcmcia, and ohci of PXA2x0.
		Ported from OpenBSD as part of zaurus port by NONAKA Kimihiro.
		[peter 20061217]
	zaurus: Add support for Sharp C3x00 PDA. [ober 20061217]
	kernel: Merge yamt-splraseipl branch, which provides splraiseipl(9)
		and makeiplcookie(9). [yamt 20061221]
	sgimips: Add support for Set Engineering's GIO Fast Ethernet board.
		[rumble 20061223]
	sgimips: Add support for the SGI Light/Entry/Starter LG1/LG2
		framebuffers found in Indigo systems. [rumble 20061226]
	nfs: Remove NQNFS support from kernel. [yamt 20061227]
	kernel, srtconfig(1): First import of the source-address-based routing
		pseudo-device.  [mouse 20061229]
	sgimips: Add i8254 timecounter(9) support for use on IP12.
		[rumble 20061229]
	mount_psshfs(8): Add puffs sshfs. [pooka 20061229]
	njata(4): Add support for Workbit CF32A CF adapter,
		aka BUFFALO RCF-CBA2.  [ichiro 20061231]
	amd64, i386: Report and work around some of the known errata for
		Athlon 64 and Opteron processors. [ad 20070101]
	i386: Add fatboot, sub-512 byte bootloader that can be written into
		the first sector of a FAT16 filesystem and will load
		the secondary bootloader from it. [dsl 20070101]
	kernel: Make mount(2) and unmount(2) use kauth(9) for security
		policy.  One incompatible change is that an unprivileged
		mount without the MNT_NOSUID and MNT_NODEV flags
		will now fail, whereas previously those flags were
		automatically added.  [elad 20070102]
	udf: Implement read-only access to UDF 2.50 and UDF 2.60 discs
		that use metadata partition mappings. [reinoud 20070104]
	i386: Add support for Microsoft Xbox. [jmcneill 20070105]
	ntp: Import ntp 4.2.4 (20061228). Tracking of network interface
		address changes is now supported - no more ntpd
		restarts when interface addresses are re-configured
		(WLAN/DSL/ppp/...). [kardel 20070106]
	config(1): Introduce a new -L option, generates a lint configuration.
		[cube 20070108]
	umass(4): Add a quirk for the Sony Cybershot USB camera.
		From Mihai Chelaru. [christos 20070109]
	kernel: Global 'securelevel' variable removed; it is now internal to
		the secmodel. [elad 20070110]
	gcc(1): Updated to version 4.1.2-20070110. [mrg 20070111]
	zaurus: Add MI todr(9) and timecounter(9) support. [ober 20070112]
	acpitools: Added from FreeBSD via fvdl. [christos 20070114]
	kernel: Add syscalls for POSIX Real-time library. [ad 20070116]
	adb(4): Add new and simplified MI ADB drivers. [macallan 20070117]
	macppc: Add a driver for the SGS 7433 mixer found in some G3 Macs.
		[macallan 20070117]
	viaenv(4): Add support for the VIA VT8231 hardware monitor and
		remove viapm(4) code. From OpenBSD. [xtraeme 20070120]
	scsictl(8): Added setspeed command, to set the highest speed that the
		optical drive should use for reading data,
		contributed by Kimura Fuyuki. [bouyer 20070122]
	msk(4): Pull support for some devices and many bug fixes from OpenBSD.
		[msaitoh 20070130]
	uplcom(4): Add support for Willcom WS002IN PHS device
		(Prolific Technology PL2303X). [msaitoh 20070131]
	zaurus: Updates to allow Zaurus screen to rotate 90 degrees to a usable
		state with the keyboard. From OpenBSD. [ober,peter 20070202]
	security(8): Add support for per-user /tmp. [elad 20070204]
	ugensa(4): Add support Novatel Merlin, used in VZ V620.
		From John Nielsen on freebsd-mobile. [gdt 20070205]
	i386, amd64: Add cmos(4), driver for x86 CMOS RAM. [dyoung 20070206]
	find(1): Add -delete option from FreeBSD. [elad 20070206]
	rum(4): Add support for MELCO WLI-U2-SG54HP and PLANEX GW-US54Mini2.
		[elad 20070206]
	kernel: Merge the newlock2 branch. Introduces a number of new
		kernel synchronization primitives, improves scalability
		on MP systems, and replaces the existing SA threads model
		with a 1:1 threading model. [ad 20070209]
	news68k: Implement generic softintr(9) support. [tsutsui 20070210]
	ubsa(4): Add support for CDMA modems sold by Eurotel (now O2).
		[salo 20070210]
	refuse: Added FUSE (file system in userspace) compatibility
		functionality, layered on top of libpuffs. [agc 20070211]
	i386: Added an opencrypto provider for VIA ACE (AES encryption
		instructions), ported from OpenBSD. [daniel 20070217]
	shark: Add generic softintr(9) support. [matt 20070218]
	kernel: Add code to support per-system call statistics. [dsl 20070218]
	sdiff(1): Replaced by OpenBSD sdiff(1) (public domain). [rmind 20070219]
	sgimips: Support the sq(4) interface on the Challenge S's IOPLUS

usr.bin/ul/ul.c
/*	$NetBSD: ul.c,v 1.13 2008/07/21 14:19:27 lukem Exp $	*/

/*
 * Copyright (c) 1980, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#ifndef lint
__COPYRIGHT("@(#) Copyright (c) 1980, 1993\
 The Regents of the University of California.  All rights reserved.");
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)ul.c	8.1 (Berkeley) 6/6/93";
#endif
__RCSID("$NetBSD: ul.c,v 1.13 2008/07/21 14:19:27 lukem Exp $");
#endif /* not lint */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <termcap.h>
#include <unistd.h>

#define	IESC	'\033'
#define	SO	'\016'
#define	SI	'\017'
#define	HFWD	'9'
#define	HREV	'8'
#define	FREV	'7'
#define	MAXBUF	512

#define	NORMAL	000
#define	ALTSET	001	/* Reverse */
#define	SUPERSC	002	/* Dim */
#define	SUBSC	004	/* Dim | Ul */
#define	UNDERL	010	/* Ul */
#define	BOLD	020	/* Bold */

struct tinfo *info;
int	must_use_uc, must_overstrike;
char	*CURS_UP, *CURS_RIGHT, *CURS_LEFT,
	*ENTER_STANDOUT, *EXIT_STANDOUT, *ENTER_UNDERLINE, *EXIT_UNDERLINE,
	*ENTER_DIM, *ENTER_BOLD, *ENTER_REVERSE, *UNDER_CHAR, *EXIT_ATTRIBUTES;

struct	CHAR	{
	char	c_mode;
	char	c_char;
} ;

struct	CHAR	obuf[MAXBUF];
int	col, maxcol;
int	mode;
int	halfpos;
int	upln;
int	iflag;
int	sflag;

int	main __P((int, char **));
void	filter __P((FILE *));
void	flushln __P((void));
void	fwd __P((void));
void	iattr __P((void));
void	initbuf __P((void));
void	initcap __P((void));
void	outc __P((int));
int	outchar __P((int));
void	overstrike __P((void));
void	reverse __P((void));
void	setulmode __P((int));


#define	PRINT(s)	if (s == NULL) /* void */; else tputs(s, 1, outchar)

int
main(argc, argv)
	int argc;
	char **argv;
{
	int c;
	char *termtype;
	FILE *f;

	termtype = getenv("TERM");
	if (termtype == NULL || (argv[0][0] == 'c' && !isatty(1)))
		termtype = "lpr";
	while ((c=getopt(argc, argv, "ist:T:")) != -1)
		switch(c) {

		case 't':
		case 'T': /* for nroff compatibility */
				termtype = optarg;
			break;
		case 'i':
			iflag = 1;
			break;
		case 's':
			sflag = 1;
			break;

		default:
			fprintf(stderr,
				"usage: %s [ -i ] [ -s ] [ -tTerm ] file...\n",
				argv[0]);
			exit(1);
		}

	switch(t_getent(&info, termtype)) {

	case 1:
		break;

	default:
		fprintf(stderr,"trouble reading termcap");
		/* fall through to ... */

	case 0:
		/* No such terminal type - assume dumb */
		if (t_setinfo(&info, "dumb:os:col#80:cr=^M:sf=^J:am:") < 0) {
			fprintf(stderr, "t_setinfo failed, cannot continue\n");
			exit(1);
		}
		
		break;
	}
	initcap();
	if (    (t_getflag(info, "os") && ENTER_BOLD==NULL ) ||
		(t_getflag(info, "ul") && ENTER_UNDERLINE==NULL
		 && UNDER_CHAR==NULL))
			must_overstrike = 1;
	initbuf();
	if (optind == argc)
		filter(stdin);
	else for (; optind<argc; optind++) {
		f = fopen(argv[optind],"r");
		if (f == NULL) {
			perror(argv[optind]);
			exit(1);
		}

		filter(f);
		fclose(f);
	}
	exit(0);
}

void
filter(f)
	FILE *f;
{
	int c;

	while ((c = getc(f)) != EOF) switch(c) {

	case '\b':
		if (col > 0)
			col--;
		continue;

	case '\t':
		col = (col+8) & ~07;
		if (col > maxcol)
			maxcol = col;
		continue;

	case '\r':
		col = 0;
		continue;

	case SO:
		mode |= ALTSET;
		continue;

	case SI:
		mode &= ~ALTSET;
		continue;

	case IESC:
		switch (c = getc(f)) {

		case HREV:
			if (halfpos == 0) {
				mode |= SUPERSC;
				halfpos--;
			} else if (halfpos > 0) {
				mode &= ~SUBSC;
				halfpos--;
			} else {
				halfpos = 0;
				reverse();
			}
			continue;

		case HFWD:
			if (halfpos == 0) {
				mode |= SUBSC;
				halfpos++;
			} else if (halfpos < 0) {
				mode &= ~SUPERSC;
				halfpos++;
			} else {
				halfpos = 0;
				fwd();
			}
			continue;

		case FREV:
			reverse();
			continue;

		default:
			fprintf(stderr,
				"Unknown escape sequence in input: %o, %o\n",
				IESC, c);
			exit(1);
		}
		continue;

	case '_':
		if (obuf[col].c_char)
			obuf[col].c_mode |= UNDERL | mode;
		else
			obuf[col].c_char = '_';
		col++;
		if (col > maxcol)
			maxcol = col;
		continue;

	case '\n':
		flushln();
		continue;

	case '\f':
		flushln();
		putchar('\f');
		continue;

	case ' ':
		if (sflag) {
			col ++;
			if (col > maxcol) maxcol = col;
			continue;
		}
		/* fall through */

	default:
		if (c < ' ')	/* non printing */
			continue;
		if (obuf[col].c_char == '\0') {
			obuf[col].c_char = c;
			obuf[col].c_mode = mode;
		} else if (obuf[col].c_char == '_') {
			obuf[col].c_char = c;
			obuf[col].c_mode |= UNDERL|mode;
		} else if (obuf[col].c_char == c)
			obuf[col].c_mode |= BOLD|mode;
		else
			obuf[col].c_mode = mode;
		col++;
		if (col > maxcol)
			maxcol = col;
		continue;
	}
	if (maxcol)
		flushln();
}

void
flushln()
{
	int lastmode;
	int i;
	int hadmodes = 0;

	lastmode = NORMAL;
	for (i=0; i<maxcol; i++) {
		if (obuf[i].c_mode != lastmode) {
			hadmodes++;
			setulmode(obuf[i].c_mode);
			lastmode = obuf[i].c_mode;
		}
		if (obuf[i].c_char == '\0') {
			if (upln) {
				PRINT(CURS_RIGHT);
			}
			else {
				outc(' ');
			}
		} else
			outc(obuf[i].c_char);
	}
	if (lastmode != NORMAL) {
		setulmode(0);
	}
	if (must_overstrike && hadmodes)
		overstrike();
	putchar('\n');
	if (iflag && hadmodes)
		iattr();
	(void)fflush(stdout);
	if (upln)
		upln--;
	initbuf();
}

/*
 * For terminals that can overstrike, overstrike underlines and bolds.
 * We don't do anything with halfline ups and downs, or Greek.
 */
void
overstrike()
{
	int i;
	char lbuf[256];
	char *cp = lbuf;
	int hadbold=0;

	/* Set up overstrike buffer */
	for (i=0; i<maxcol; i++)
		switch (obuf[i].c_mode) {
		case NORMAL:
		default:
			*cp++ = ' ';
			break;
		case UNDERL:
			*cp++ = '_';
			break;
		case BOLD:
			*cp++ = obuf[i].c_char;
			hadbold=1;
			break;
		}
	putchar('\r');
	for (*cp=' '; *cp==' '; cp--)
		*cp = 0;
	for (cp=lbuf; *cp; cp++)
		putchar(*cp);
	if (hadbold) {
		putchar('\r');
		for (cp=lbuf; *cp; cp++)
			putchar(*cp=='_' ? ' ' : *cp);
		putchar('\r');
		for (cp=lbuf; *cp; cp++)
			putchar(*cp=='_' ? ' ' : *cp);
	}
}

void
iattr()
{
	int i;
	char lbuf[256];
	char *cp = lbuf;

	for (i=0; i<maxcol; i++)
		switch (obuf[i].c_mode) {
		case NORMAL:	*cp++ = ' '; break;
		case ALTSET:	*cp++ = 'g'; break;
		case SUPERSC:	*cp++ = '^'; break;
		case SUBSC:	*cp++ = 'v'; break;
		case UNDERL:	*cp++ = '_'; break;
		case BOLD:	*cp++ = '!'; break;
		default:	*cp++ = 'X'; break;
		}
	for (*cp=' '; *cp==' '; cp--)
		*cp = 0;
	for (cp=lbuf; *cp; cp++)
		putchar(*cp);
	putchar('\n');
}

void
initbuf()
{

	memset((char *)obuf, 0, sizeof (obuf));	/* depends on NORMAL == 0 */
	col = 0;
	maxcol = 0;
	mode &= ALTSET;
}

void
fwd()
{
	int oldcol, oldmax;

	oldcol = col;
	oldmax = maxcol;
	flushln();
	col = oldcol;
	maxcol = oldmax;
}

void
reverse()
{
	upln++;
	fwd();
	PRINT(CURS_UP);
	PRINT(CURS_UP);
	upln++;
}

void
initcap()
{
	/* This nonsense attempts to work with both old and new termcap */
	CURS_UP =		t_agetstr(info, "up");
	CURS_RIGHT =		t_agetstr(info, "ri");
	if (CURS_RIGHT == NULL)
		CURS_RIGHT =	t_agetstr(info, "nd");
	CURS_LEFT =		t_agetstr(info, "le");
	if (CURS_LEFT == NULL)
		CURS_LEFT =	t_agetstr(info, "bc");
	if (CURS_LEFT == NULL && t_getflag(info, "bs"))
		CURS_LEFT =	"\b";

	ENTER_STANDOUT =	t_agetstr(info, "so");
	EXIT_STANDOUT =		t_agetstr(info, "se");
	ENTER_UNDERLINE =	t_agetstr(info, "us");
	EXIT_UNDERLINE =	t_agetstr(info, "ue");
	ENTER_DIM =		t_agetstr(info, "mh");
	ENTER_BOLD =		t_agetstr(info, "md");
	ENTER_REVERSE =		t_agetstr(info, "mr");
	EXIT_ATTRIBUTES =	t_agetstr(info, "me");

	if (!ENTER_BOLD && ENTER_REVERSE)
		ENTER_BOLD = ENTER_REVERSE;
	if (!ENTER_BOLD && ENTER_STANDOUT)
		ENTER_BOLD = ENTER_STANDOUT;
	if (!ENTER_UNDERLINE && ENTER_STANDOUT) {
		ENTER_UNDERLINE = ENTER_STANDOUT;
		EXIT_UNDERLINE = EXIT_STANDOUT;
	}
	if (!ENTER_DIM && ENTER_STANDOUT)
		ENTER_DIM = ENTER_STANDOUT;
	if (!ENTER_REVERSE && ENTER_STANDOUT)
		ENTER_REVERSE = ENTER_STANDOUT;
	if (!EXIT_ATTRIBUTES && EXIT_STANDOUT)
		EXIT_ATTRIBUTES = EXIT_STANDOUT;
	
	/*
	 * Note that we use REVERSE for the alternate character set,
	 * not the as/ae capabilities.  This is because we are modelling
	 * the model 37 teletype (since that's what nroff outputs) and
	 * the typical as/ae is more of a graphics set, not the greek
	 * letters the 37 has.
	 */

	UNDER_CHAR =		t_agetstr(info, "uc");
	must_use_uc = (UNDER_CHAR && !ENTER_UNDERLINE);
}

int
outchar(c)
	int c;
{
	return (putchar(c & 0177));
}

static int curmode = 0;

void
outc(c)
	int c;
{
	putchar(c);
	if (must_use_uc && (curmode&UNDERL)) {
		PRINT(CURS_LEFT);
		PRINT(UNDER_CHAR);
	}
}

void
setulmode(newmode)
	int newmode;
{
	if (!iflag) {
		if (curmode != NORMAL && newmode != NORMAL)
			setulmode(NORMAL);
		switch (newmode) {
		case NORMAL:
			switch(curmode) {
			case NORMAL:
				break;
			case UNDERL:
				PRINT(EXIT_UNDERLINE);
				break;
			default:
				/* This includes standout */
				PRINT(EXIT_ATTRIBUTES);
				break;
			}
			break;
		case ALTSET:
			PRINT(ENTER_REVERSE);
			break;
		case SUPERSC:
			/*
			 * This only works on a few terminals.
			 * It should be fixed.
			 */
			PRINT(ENTER_UNDERLINE);
			PRINT(ENTER_DIM);
			break;
		case SUBSC:
			PRINT(ENTER_DIM);
			break;
		case UNDERL:
			PRINT(ENTER_UNDERLINE);
			break;
		case BOLD:
			PRINT(ENTER_BOLD);
			break;
		default:
			/*
			 * We should have some provision here for multiple modes
			 * on at once.  This will have to come later.
			 */
			PRINT(ENTER_STANDOUT);
			break;
		}
	}
	curmode = newmode;
}
